apps/snitch_core/lib/core/domain/shipment_engine.ex

```diff
@@ -0,0 +1,161 @@
+
+  Support for `shipments` partially fulfilling an `order` is yet to be added.
+
+  Constraints to be added in future:
+    * prefer shorter shipping distance.
+    * prefer smaller shipping-cost (to user)
+    * prefer smaller shipping cost (to store owner)
+
+  > Note
+  > - Partial fulfillment refers to a scenario when a lineitem is fulfilled 
+    by two different shipments. 
+  > - The module doesn't take into consideration partial fulfillment and backorders.
+  """
+
+  alias Aruspex.Problem
+  alias Snitch.Data.Schema.Order
+  alias Aruspex.Strategy.SimulatedAnnealing
+
+  @domain [true, false]
+
+  @doc """
+  Returns the optimal shipment.
+
+  Takes a list of `packages` and an `order` struct.
+  An optimal shipment which fulfills the specified business
+  constraints is returned.
+  At present `shipments` which fulfill the `order` and are mutually
+  exclusive in terms of `lineitem` fulfillment are being returned.
+  """
+  @spec run(list, Order.t()) :: list
+  def run(packages, %Order{} = order) do
+    packages_with_id = append_keys(packages)
+    edges = create_csp(packages_with_id)
+
+    vars =
+      Enum.reduce(packages_with_id, [], fn p, acc ->
+        [p.id | acc]
+      end)
+
+    item_var_map = item_var_mapping(packages_with_id)
+
+    problem = Problem.new()
+
+    for var <- vars, do: Problem.add_variable(problem, var, @domain)
+
+    binary_constraint(problem, edges)
+    summation_constraint(problem, vars, item_var_map, order)
+
+    result =
+      problem
+      |> SimulatedAnnealing.set_strategy()
+      |> Enum.take(1)
+
+    bindings = variable_assignment(result)
+    filter_packages(item_var_map, bindings)
+  end
+
+  defp variable_assignment([]), do: []
+
+  defp variable_assignment([result]) do
+    Enum.reject(result.binding, fn
+      {{:hidden, _, _}, _} -> true
+      _ -> false
+    end)
+  end
+
+  defp filter_packages(pckgs, bindings) do
+    Enum.reduce(bindings, [], fn
+      {id, true}, acc -> [pckgs[id] | acc]
+      {_, false}, acc -> acc
+    end)
+  end
+
+  defp binary_constraint(problem, edges) do
+    Enum.map(edges, fn {x, y} ->
+      Problem.post(problem, x.id, y.id, &(not (&1 and &2)))
+    end)
+  end
+
+  defp summation_constraint(problem, vars, item_var_map, order) do
+    Problem.post(problem, vars, fn values ->
+      item_count =
+        vars
+        |> Stream.zip(values)
+        |> Enum.reduce(0, fn
+          {var, true}, acc -> length(item_var_map[var].items) + acc
+          {_, false}, acc -> acc
+        end)
+
+      item_count == length(order.line_items)
+    end)
+  end
+
+  defp item_var_mapping(packages) do
+    Enum.map(packages, fn pkg ->
+      {pkg.id, pkg}
+    end)
+  end
+
+  defp create_csp(packages) do
+    packages
+    |> append_keys()
+    |> Enum.map(&group_package_with_item(&1))
+    |> List.flatten()
+    |> group_packages_by_item()
+    |> packages_with_same_item()
+    |> find_unique_edges()
+  end
+
+  defp append_keys(packages) do
```

Comment:
So what if you're building atoms? isn't the "reduction" on the list of integers?
Here's how I would use `with_index`:
```elixir
packages
|> Stream.with_index()
|> Enum.map(fn {package, index} ->
  Map.put(package, :id, String.to_atom("p#{index}"))
end)
```